amethyst_tiles/src/pass.rs

```diff
@@ -0,0 +1,503 @@
+    for DrawTiles2DDesc<T, E>
+{
+    fn build(
+        self,
+        _ctx: &GraphContext<B>,
+        factory: &mut Factory<B>,
+        _queue: QueueId,
+        _aux: &Resources,
+        framebuffer_width: u32,
+        framebuffer_height: u32,
+        subpass: hal::pass::Subpass<'_, B>,
+        _buffers: Vec<NodeBuffer>,
+        _images: Vec<NodeImage>,
+    ) -> Result<Box<dyn RenderGroup<B, Resources>>, failure::Error> {
+        #[cfg(feature = "profiler")]
+        profile_scope!("build");
+
+        let env = FlatEnvironmentSub::new(factory)?;
+        let textures = TextureSub::new(factory)?;
+        let vertex = DynamicVertexBuffer::new();
+
+        let (pipeline, pipeline_layout) = build_tiles_pipeline(
+            factory,
+            subpass,
+            framebuffer_width,
+            framebuffer_height,
+            true,
+            vec![env.raw_layout(), textures.raw_layout()],
+        )?;
+
+        Ok(Box::new(DrawTiles2D::<B, T, E> {
+            pipeline,
+            pipeline_layout,
+            env,
+            textures,
+            vertex,
+            sprites: Default::default(),
+            _marker: PhantomData::default(),
+        }))
+    }
+}
+
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+pub struct DrawTiles2D<B: Backend, T: Tile, E: CoordinateEncoder> {
+    pipeline: B::GraphicsPipeline,
+    pipeline_layout: B::PipelineLayout,
+    env: FlatEnvironmentSub<B>,
+    textures: TextureSub<B>,
+    vertex: DynamicVertexBuffer<B, SpriteArgs>,
+    sprites: OneLevelBatch<TextureId, SpriteArgs>,
+
+    #[derivative(Debug = "ignore")]
+    _marker: PhantomData<(T, E)>,
+}
+
+impl<B: Backend, T: Tile, E: CoordinateEncoder> RenderGroup<B, Resources> for DrawTiles2D<B, T, E> {
+    fn prepare(
+        &mut self,
+        factory: &Factory<B>,
+        _queue: QueueId,
+        index: usize,
+        _subpass: hal::pass::Subpass<'_, B>,
+        resources: &Resources,
+    ) -> PrepareResult {
+        #[cfg(feature = "profiler")]
+        profile_scope!("prepare");
+
+        let (sprite_sheet_storage, tex_storage, hiddens, tile_maps) = <(
+            Read<'_, AssetStorage<SpriteSheet>>,
+            Read<'_, AssetStorage<Texture>>,
+            ReadStorage<'_, Hidden>,
+            ReadStorage<'_, TileMap<T, E>>,
+        )>::fetch(resources);
+        let sprites_ref = &mut self.sprites;
+        let textures_ref = &mut self.textures;
+
+        sprites_ref.clear_inner();
+
+        self.env.process(factory, index, resources);
+
+        if let Some((tile_map, _)) = (&tile_maps, !&hiddens).join().next() {
+            // Render the first tilemap we find
+            let (top, mut bottom) = fetch_world_bounds(tile_map, resources);
+            bottom.z = 0;
+
+            let camera_region = Region::new(top, bottom);
+            camera_region
+                .iter()
+                .filter_map(|coord| {
+                    if let Some(tile) = tile_map.get(&coord) {
```

Comment:
This deep nesting on `Option` inside iterator is a bit disorienting here.

How about some early return?
```rust
let tile = tile_map.get(&coord)?;
let sprite_number = tile.sprite(resources)?;
```